prac_07/project_management.py

```diff
@@ -0,0 +1,233 @@
+            for project in new_projects:
+                projects.append(project)
+        elif choice == "S":
+            out_filename = get_valid_filename()
+            save_projects(projects, out_filename)
+        elif choice == "D":
+            projects.sort()  # Sorted by priority
+            display_projects(projects)
+        elif choice == "F":
+            filter_date_string = get_valid_date_string()
+            filter_projects_by_date(projects, filter_date_string)
+        elif choice == "A":
+            name, start_date, priority, cost_estimate, completion_percentage = get_valid_project_details()
+            projects.append(Project(name, start_date, priority, cost_estimate, completion_percentage))
+        elif choice == "U":
+            update_project(projects)
+        else:
+            print("Invalid choice!")
+        print(MENU)
+        choice = input(">>> ").upper()
+    save_projects(projects, DEFAULT_FILENAME)
+    print("Finished.")
+
+
+def get_valid_filename():
+    """Get a valid filename with the right file extension."""
+    is_valid_filename = False
+    while not is_valid_filename:
+        part = input("Filename (.txt): ")
+        if part == "":
+            print("Input can not be empty!")
+        elif ".txt" not in part:
+            print("Must have the right file extension!")
+        else:
+            is_valid_filename = True
+
+    return part  # No problem with potential undefined variable
+
+
+def load_projects(filename):
+    """Load projects from file."""
+    projects = []
+    try:
+        with open(filename, 'r') as in_file:
+            in_file.readline()  # Ignore header
+            for line in in_file:
+                parts = line.strip().split('\t')
+                name = parts[0]
+                start_date = parts[1]
+                priority = int(parts[2])
+                cost_estimate = float(parts[3])
+                completion_percentage = float(parts[4])
+                projects.append(Project(name, start_date, priority, cost_estimate, completion_percentage))
+
+            print(f"{len(projects)} projects loaded.")
+
+    except ValueError:
+        print("Make sure data in the file matches the format!\n"
+              "e.g. Name	Start Date	Priority	Cost Estimate	Completion Percentage")
+
+    except FileNotFoundError:
+        print("File not found!")
+
+    return projects
+
+
+def save_projects(projects, filename):
+    """Save projects into file."""
+    with open(filename, 'w') as out_file:
+        print("Name\tStart Date\tPriority\tCost Estimate\tCompletion Percentage", file=out_file)
+        for project in projects:
+            print(f"{project.name}\t{project.start_date}\t{project.priority}\t{project.cost_estimate}\t"
+                  f"{project.completion_percentage}", file=out_file)
+
+
+def display_projects(projects):
+    """Display projects."""
+    incomplete_projects = []
+    completed_projects = []
+
+    for project in projects:
```

Comment:
I see that you sort the projects outside the function, should you have this detail within the function?

Reply:
I think I should! 🙈